monitor/src/main/java/edu/asu/diging/monitor/web/admin/AddAppController.java

```diff
@@ -39,14 +54,39 @@ public String show(Model model) {
             recipient.setEmail(r.getEmail());
             return recipient;
         }).collect(Collectors.toList()));
+        appForm.setGroupIds(groupManager.getGroups().stream()
+                .collect(Collectors.toMap(g -> g.getId(), g -> g.getName())).entrySet());
         return "admin/apps/show";
     }
-	
-	@RequestMapping(value="/admin/apps/add", method=RequestMethod.POST)
-	public String add(@ModelAttribute AppForm appForm) {
-		IApp app = new App();
-		appHelper.copyAppInfo(app, appForm);
-		appManager.addApp(app);
-		return "redirect:/admin/apps/add";
-	}
+
+    @RequestMapping(value = "/admin/apps/add", method = RequestMethod.POST)
+    public String add(@ModelAttribute AppForm appForm, RedirectAttributes redirectAttrs) {
+        IApp app = new App();
+        if (!groupHelper.isValid(appForm, redirectAttrs)) {
```

Comment:
this should use a validator (see e.g. https://www.boraji.com/spring-4-mvc-form-validation-example-using-validator-interface, but there are plenty out there)